### PR TITLE
Convert eviction code to STL

### PR DESCRIPTION
**Pull request by pcmoritz**
After this PR, all the utarray/uthash code will be removed from the plasma store and we can modernize the code some more.

### CODE REVIEW

**Review by robertnishihara — COMMENTED** [review]
Nice job! Our test coverage of eviction is pretty minimal, so I'll try it out a little to see if it seems to be doing the right thing.

File: src/plasma/test/run_tests.sh [hunk 1]
```diff
@@ -13,7 +13,7 @@ killall plasma_store
 redis_pid=$!
 sleep 1
 # flush the redis server
-../common/thirdparty/redis/src/redis-cli flushall &
+./src/common/thirdparty/redis/src/redis-cli flushall &
```
Comment (pcmoritz) [inline src/plasma/test/run_tests.sh]: No, but this one is more uniform (look at the code)

File: src/plasma/eviction_policy.cc [hunk 2]
```diff
@@ -112,47 +78,33 @@ int64_t EvictionState_choose_objects_to_evict(
...
-        break;
-      }
-      (*objects_to_evict)[counter] = element->object_id;
-      /* Update the LRU cache. */
-      remove_object_from_lru_cache(eviction_state, element->object_id);
-      counter += 1;
-    }
+    int64_t result_size = objs_to_evict.size() * sizeof(ObjectID);
+    *objects_to_evict = (ObjectID *) malloc(result_size);
+    memcpy(*objects_to_evict, objs_to_evict.data(), result_size);
```
Comment (pcmoritz) [inline src/plasma/eviction_policy.cc]: This is valid use of vectors, but I'll remove it in the next big PR, see https://github.com/ray-project/ray/pull/535

File: src/plasma/eviction_policy.cc [hunk 3]
```diff
@@ -112,47 +78,33 @@ int64_t EvictionState_choose_objects_to_evict(
     int64_t num_bytes_required,
     int64_t *num_objects_to_evict,
     ObjectID **objects_to_evict) {
```
Comment (pcmoritz) [inline src/plasma/eviction_policy.cc]: Yes, I'm doing that in https://github.com/ray-project/ray/pull/535

File: src/plasma/eviction_policy.cc [hunk 4]
```diff
@@ -1,109 +1,75 @@
 #include "eviction_policy.h"
 
 #include "utlist.h"
+#include "uthash.h"
```
Comment (pcmoritz) [inline src/plasma/eviction_policy.cc]: originally done in https://github.com/ray-project/ray/pull/535, I moved it

File: src/plasma/eviction_policy.cc [hunk 5]
```diff
@@ -1,109 +1,75 @@
...
-  ReleasedObject *released_object;
-  /** Handle for the uthash table. */
-  UT_hash_handle handle;
-} released_object_entry;
+#include <list>
+#include <unordered_map>
+
+class LRUCache {
+ private:
+  // A doubly-linked list containing the items in the cache and
```
Comment (pcmoritz) [inline src/plasma/eviction_policy.cc]: cone

File: src/plasma/eviction_policy.cc [hunk 6]
```diff
@@ -1,109 +1,75 @@
...
+  // their sizes in LRU order.
+  std::list<std::pair<ObjectID, int64_t>> item_list_;
+  // A hash table mapping the object ID of objects in the cache to its location
```
Comment (pcmoritz) [inline src/plasma/eviction_policy.cc]: done

File: src/plasma/eviction_policy.cc [hunk 7]
```diff
@@ -1,109 +1,75 @@
...
+  // in the doubly linked list item_list_.
+  std::unordered_map<ObjectID, decltype(item_list_.begin()), UniqueIDHasher>
```
Comment (pcmoritz) [inline src/plasma/eviction_policy.cc]: done

File: src/plasma/eviction_policy.cc [hunk 8]
```diff
@@ -1,109 +1,75 @@
...
+      item_map_;
+
+ public:
+  LRUCache(){};
```
Comment (pcmoritz) [inline src/plasma/eviction_policy.cc]: Clang changed it to this. I'm also surprised.

File: src/plasma/eviction_policy.cc [hunk 9]
```diff
@@ -1,109 +1,75 @@
...
+
+  void store(const ObjectID &key, int64_t size) {
```
Comment (pcmoritz) [inline src/plasma/eviction_policy.cc]: yeah

File: src/plasma/eviction_policy.cc [hunk 10]
```diff
@@ -1,109 +1,75 @@
...
+    auto it = item_map_.find(key);
+    if (it == item_map_.end()) {
```
Comment (pcmoritz) [inline src/plasma/eviction_policy.cc]: done

File: src/plasma/eviction_policy.cc [hunk 11]
```diff
@@ -1,109 +1,75 @@
...
+      item_list_.push_front(std::make_pair(key, size));
+      item_map_.insert(std::make_pair(key, item_list_.begin()));
+    }
+  }
+
+  void remove(const ObjectID &key) {
+    auto it = item_map_.find(key);
+    if (it != item_map_.end()) {
```
Comment (pcmoritz) [inline src/plasma/eviction_policy.cc]: done

File: src/plasma/eviction_policy.cc [hunk 12]
```diff
@@ -1,109 +1,75 @@
...
+      item_list_.erase(it->second);
+      item_map_.erase(it);
+    }
+  }
+
+  void access(const ObjectID &key) {
```
Comment (pcmoritz) [inline src/plasma/eviction_policy.cc]: It brings the object to the front of the cache. I want to do some refactoring in the future that uses this, for now I removed it.

File: src/plasma/eviction_policy.cc [hunk 13]
```diff
@@ -1,109 +1,75 @@
...
+    auto it = item_map_.find(key);
+    CHECK(it != item_map_.end());
+    item_list_.splice(item_list_.begin(), item_list_, it->second);
+  }
+
+  int64_t choose_objects_to_evict(int64_t num_bytes_required,
+                                  std::vector<ObjectID> &objects_to_evict) {
+    int64_t bytes_evicted = 0;
+    auto it = item_list_.end();
+    while (bytes_evicted < num_bytes_required && it != item_list_.begin()) {
```
Comment (pcmoritz) [inline src/plasma/eviction_policy.cc]: It is more natural to me to have the hottest item at the front ;)

**Comment by robertnishihara** [comment]
Looks like the test failure is #483.

**Pull request merged**